File: Router/server/src/services/rag/retriever.js
import { charNgrams, msSince, clipText } from "./textUtils.js";
import { loadFaqItemsFromRows, buildItemMap, itemToResult } from "./dataLoader.js";
import { readIndexMeta } from "./indexStatus.js";
import { ragKbAssetsDirPath } from "../paths.js";
import { SEARCHABLE_DOC_TYPES } from "./searchDocBuilder.js";
import { aggregateTokens } from "./tokenUtils.js";
import { RagLogSink, formatRagSearchSummary } from "./ragLogger.js";

/**
 * RAG 检索器 — 向量 + 关键词混合检索、RRF 融合、rerank、chat 直出/合成。
 * 调试页 POST /rag/chat 的核心执行类。
 */
export class RagRetriever {
    settings;
    kbId;
    ragModelsStore;
    weaviateStore;
    embeddingClient;
    ragLlmClient;
    runtimeConfig;
    itemMap = null;
    log;
    ragCtx;

    constructor(kbId, ctx, runtimeConfig, logModule = "rag-debug") {
        this.kbId = kbId;
        this.ragCtx = ctx;
        this.settings = ctx.settings;
        this.ragModelsStore = ctx.ragModelsStore;
        this.weaviateStore = ctx.weaviateStore;
        this.embeddingClient = ctx.embeddingClient;
        this.ragLlmClient = ctx.ragLlmClient;
        this.runtimeConfig = runtimeConfig;
        this.log = new RagLogSink(ctx.opLog, logModule, kbId);
    }

    rt(name, defaultVal) {
        if (this.runtimeConfig && this.runtimeConfig[name] != null)
            return this.runtimeConfig[name];
        return defaultVal;
    }

    async loadItems() {
        if (this.itemMap)
            return this.itemMap;
        const store = this.ragCtx.getRagQuestionsStore(this.kbId);
        const doc = await store.getDocument();
        const assetsDir = ragKbAssetsDirPath(this.settings.filesRoot, this.kbId);
        this.itemMap = buildItemMap(loadFaqItemsFromRows(doc.items, assetsDir));
        return this.itemMap;
    }

    /**
     * 混合检索：embed query → Weaviate 向量检索 → 关键词 n-gram → RRF 融合 → rerank。
     * 例：query="怎么调光圈" 可能命中 q012 光圈调节相关 FAQ。
     */
    async search(query, topK) {
        const top_k = topK ?? this.rt("top_k", 8);
        const timing = {};
        const tTotal = performance.now();
        this.log.log(`[search] 开始 query="${clipText(query, 120)}" top_k=${top_k} use_rerank=${this.rt("use_rerank", true)}`, "step", "search");

        const tEmb = performance.now();
        const { vector: qVec, usage: embUsage } = await this.embeddingClient.embedQuery(query);
        timing.embedding_ms = msSince(tEmb);
        this.log.timing("embedding", timing.embedding_ms);
        if (embUsage)
            this.log.log(`[search] embedding tokens prompt=${embUsage.prompt_tokens ?? 0} total=${embUsage.total_tokens ?? 0}`, "step", "embedding");
        const tokenBreakdown = [];
        if (embUsage)
            tokenBreakdown.push({ phase: "embedding", usage: embUsage });

        const tVec = performance.now();
        const vectorHits = this._filterSearchHits(await this.weaviateStore.hybridSearch(
            this.kbId, query, qVec, this.settings.ragVectorTopK, 0.75,
        ));
        timing.vector_lookup_ms = msSince(tVec);
        this.log.log(`[search] 向量检索 ${vectorHits.length} 条 (top ${this.settings.ragVectorTopK})`, "step", "vector");
        if (vectorHits.length)
            this.log.log(`[search] 向量 Top3: ${vectorHits.slice(0, 3).map((h, i) => `#${i + 1} ${h.item_id} score=${Number(h.score || 0).toFixed(3)}`).join(" | ")}`, "step", "vector-top");

        const tKw = performance.now();
        const keywordHits = this._filterSearchHits(await this._keywordSearch(query, this.settings.ragKeywordTopK));
        timing.keyword_search_ms = msSince(tKw);
        this.log.log(`[search] 关键词检索 ${keywordHits.length} 条`, "step", "keyword");
        if (keywordHits.length)
            this.log.log(`[search] 关键词 Top3: ${keywordHits.slice(0, 3).map((h, i) => `#${i + 1} ${h.item_id} score=${Number(h.score || 0).toFixed(3)}`).join(" | ")}`, "step", "keyword-top");

        const tFus = performance.now();
        const candidates = this._fuseAndGroup(vectorHits, keywordHits);
        timing.fusion_ms = msSince(tFus);
        this.log.log(`[search] RRF 融合 ${candidates.length} 候选 (rrf_k=${this.settings.ragRrfK})`, "step", "fusion");
        if (candidates.length)
            this.log.log(`[search] 融合 Top3: ${candidates.slice(0, 3).map((c, i) => `#${i + 1} ${c.item_id} rrf=${c.rrf_score.toFixed(4)} vec=${c.vector_score.toFixed(3)} kw=${c.keyword_score.toFixed(3)}`).join(" | ")}`, "step", "fusion-top");

        let results;
        if (!this.rt("use_rerank", true)) {
            candidates.sort((a, b) => b.rrf_score - a.rrf_score);
            timing.rerank_ms = 0;
            results = await Promise.all(candidates.slice(0, top_k).map((c) => this._candidateToResult(c)));
            this.log.log("[search] rerank 已关闭，按 RRF 排序", "step", "rerank-skip");
        }
        else {
            const tRr = performance.now();
            const { output, usage: rerankUsage } = await this._rerank(query, candidates, Math.max(top_k, 8));
            timing.rerank_ms = msSince(tRr);
            results = await Promise.all(output.slice(0, top_k).map((c) => this._candidateToResult(c)));
            this.log.timing("rerank", timing.rerank_ms);
            this.log.log(`[search] rerank 完成，返回 ${results.length} 条`, "step", "rerank");
            if (rerankUsage)
                tokenBreakdown.push({ phase: "rerank", usage: rerankUsage });
        }

        timing.search_ms = msSince(tTotal);
        timing.total_ms = timing.search_ms;
        const { tokens, token_breakdown } = aggregateTokens(tokenBreakdown);
        this.log.timing("search_total", timing.search_ms);
        this.log.log(`[search] 完成 ${formatRagSearchSummary(results)}`, "result", "search-done");
        return { results, timing, tokens, token_breakdown };
    }

    /**
     * RAG 问答主流程：先 search，再按置信度与 answer_mode 决定返回方式。
     * - no_high_confidence：Top1 分数 < min_confidence_score
     * - direct：直出 Top1 的 answer（默认）
     * - generated：用 RAG LLM 根据 sources 合成回答
     */
    async chat(query, { topN, useLlmAnswer } = {}) {
        const tTotal = performance.now();
        const top_n = topN ?? this.rt("top_n", 3);
        const rtMode = this.rt("answer_mode", "direct");
        const minConf = this.rt("min_confidence_score", 0.05);
        this.log.log(`[chat] 开始 query="${clipText(query, 120)}" top_n=${top_n} answer_mode=${rtMode} min_confidence=${minConf}`, "step", "chat");

        const { results, timing, token_breakdown: searchBreakdown } = await this.search(query, Math.max(top_n, 8));
        const tokenBreakdown = [...(searchBreakdown || [])];
        const topScore = results.length
            ? Number(results[0].rerank_score || results[0].rrf_score || 0)
            : 0;
        const highConf = results.length > 0 && topScore >= minConf;
        this.log.log(`[chat] 置信判定 top_score=${topScore.toFixed(4)} min=${minConf} high_conf=${highConf}`, "step", "confidence");

        if (!highConf) {
            timing.generate_ms = 0;
            timing.total_ms = msSince(tTotal);
            const { tokens, token_breakdown } = aggregateTokens(tokenBreakdown);
            this.log.log("[chat] 未达置信阈值，返回低置信提示", "result", "chat-low-conf");
            return {
                answer: "未找到高置信答案。你可以换一种问法，或查看下方候选结果。",
                confidence: 0,
                sources: results.slice(0, top_n),
                images: [],
                mode: "no_high_confidence",
                timing,
                tokens,
                token_breakdown,
            };
        }

        const sources = results.slice(0, top_n);
        const shouldGenerate = rtMode === "generated" || Boolean(useLlmAnswer);
        let answer;
        let mode;
        if (shouldGenerate) {
            const tGen = performance.now();
            this.log.log(`[chat] LLM 合成回答，来源 ${sources.length} 条: ${sources.map((s) => s.id).join(", ")}`, "step", "generate");
            const gen = await this.ragLlmClient.generateAnswer(query, sources, this.runtimeConfig);
            answer = gen.content;
            if (gen.usage) {
                tokenBreakdown.push({ phase: "generate", usage: gen.usage });
                this.log.log(`[chat] 生成 tokens prompt=${gen.usage.prompt_tokens ?? 0} completion=${gen.usage.completion_tokens ?? 0}`, "step", "generate-tokens");
            }
            timing.generate_ms = msSince(tGen);
            mode = "generated";
            this.log.timing("generate", timing.generate_ms);
            this.log.log(`[chat] 合成回答 ${clipText(answer, 200)}`, "step", "generate-done");
        }
        else {
            answer = sources[0].answer;
            timing.generate_ms = 0;
            mode = "direct";
            this.log.log(`[chat] 直出模式 top1=${sources[0]?.id} answer_len=${(answer || "").length}`, "step", "direct");
        }

        timing.total_ms = msSince(tTotal);
        const { tokens, token_breakdown } = aggregateTokens(tokenBreakdown);
        this.log.timing("chat_total", timing.total_ms);
        this.log.log(`[chat] 完成 mode=${mode} confidence=${topScore.toFixed(4)} sources=${sources.length}`, "result", "chat-done");
        return {
            answer,
            confidence: topScore,
            sources,
            images: this._dedupeImages(sources),
            mode,
            timing,
            tokens,
            token_breakdown,
        };
    }

    async _keywordSearch(query, limit) {
        const meta = await readIndexMeta(this.settings.filesRoot, this.kbId);
        const index = meta?.keyword_index ?? [];
        const qTokens = [...new Set(charNgrams(query, 2, 3))].slice(0, 48);
        const scored = [];
        for (const row of index) {
            if (row.is_eval_holdout)
                continue;
            if (!SEARCHABLE_DOC_TYPES.has(String(row.doc_type || "")))
                continue;
            const text = String(row.keyword_text || "");
            let score = 0;
            for (const tok of qTokens) {
                if (tok && text.includes(tok))
                    score++;
            }
            if (score > 0) {
                scored.push({
                    doc_id: row.doc_id,
                    item_id: row.item_id,
                    doc_type: row.doc_type,
                    text: row.text,
                    keyword_text: row.keyword_text,
                    score,
                    rank_source: "keyword",
                });
            }
        }
        scored.sort((a, b) => b.score - a.score);
        if (scored.length)
            return scored.slice(0, limit);

        const kwHits = await this.weaviateStore.hybridSearch(
            this.kbId, query, null, limit, 0,
        );
        return kwHits.map((h) => ({ ...h, rank_source: "keyword" }));
    }

    _fuseAndGroup(vectorHits, keywordHits) {
        const byItem = new Map();
        const rrfK = this.settings.ragRrfK;

        const add = (hit, rank, source) => {
            const itemId = String(hit.item_id);
            let cand = byItem.get(itemId);
            if (!cand) {
                cand = {
                    item_id: itemId,
                    vector_score: 0,
                    keyword_score: 0,
                    rrf_score: 0,
                    rerank_score: 0,
                    matched_doc_types: new Set(),
                    matched_doc_ids: [],
                };
                byItem.set(itemId, cand);
            }
            cand.rrf_score += 1 / (rrfK + rank);
            cand.matched_doc_types.add(String(hit.doc_type));
            const docId = String(hit.doc_id);
            if (!cand.matched_doc_ids.includes(docId))
                cand.matched_doc_ids.push(docId);
            const score = Number(hit.score || 0);
            if (source === "vector")
                cand.vector_score = Math.max(cand.vector_score, score);
            else
                cand.keyword_score = Math.max(cand.keyword_score, score);
        };

        vectorHits.forEach((h, i) => add(h, i + 1, "vector"));
        keywordHits.forEach((h, i) => add(h, i + 1, "keyword"));

        return [...byItem.values()].sort((a, b) => b.rrf_score - a.rrf_score);
    }

    _filterSearchHits(hits) {
        return hits.filter((h) => SEARCHABLE_DOC_TYPES.has(String(h.doc_type || "")));
    }

    async _rerank(query, candidates, topK) {
        const items = await this.loadItems();
        const docs = [];
        const kept = [];
        for (const cand of candidates.slice(0, Math.max(topK * 3, topK))) {
            const item = items.get(cand.item_id);
            if (!item)
                continue;
            const variants = (item.variants || []).slice(0, 8).join("；");
            docs.push(`主问题：${item.question}\n相似问法：${variants}\n答案摘要：${item.answer_summary}`);
            kept.push(cand);
        }
        if (!kept.length)
            return { output: [], usage: null };
        const { ranked, usage } = await this.ragLlmClient.rerank(query, docs, Math.min(topK, kept.length));
        const output = [];
        const used = new Set();
        for (const [idx, score] of ranked) {
            if (idx < 0 || idx >= kept.length || used.has(idx))
                continue;
            kept[idx].rerank_score = Number(score);
            output.push(kept[idx]);
            used.add(idx);
        }
        if (output.length < Math.min(topK, kept.length)) {
            for (let idx = 0; idx < kept.length; idx++) {
                if (!used.has(idx)) {
                    kept[idx].rerank_score = kept[idx].rrf_score;
                    output.push(kept[idx]);
                }
                if (output.length >= topK)
                    break;
            }
        }
        return { output, usage };
    }

    async _candidateToResult(cand) {
        const items = await this.loadItems();
        const item = items.get(cand.item_id);
        if (!item)
            return { id: cand.item_id };
        return itemToResult(item, {
            vector_score: cand.vector_score,
            keyword_score: cand.keyword_score,
            rrf_score: cand.rrf_score,
            rerank_score: cand.rerank_score,
            matched_doc_types: [...cand.matched_doc_types].sort(),
            matched_doc_ids: cand.matched_doc_ids,
        });
    }

    _dedupeImages(sources) {
        const seen = new Set();
        const images = [];
        for (const src of sources) {
            for (const img of src.images || []) {
                const key = img.src || img.url || "";
                if (key && !seen.has(key)) {
                    seen.add(key);
                    images.push(img);
                }
            }
        }
        return images;
    }
}

File: faq/app/retriever.py
from __future__ import annotations

import json
import sqlite3
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import faiss
import numpy as np

from .config import Settings, settings
from .data_loader import data_hash
from .database import connect, fetch_item, get_meta
from .embedding import EmbeddingClient
from .indexer import index_files_exist
from .llm import ModelClient
from .text_utils import char_ngrams, clip_text, keyword_text, strip_markdown
from .timing import ms_since


@dataclass
class ParentCandidate:
    item_id: str
    vector_score: float = 0.0
    keyword_score: float = 0.0
    rrf_score: float = 0.0
    rerank_score: float = 0.0
    matched_doc_types: set[str] = field(default_factory=set)
    matched_doc_ids: list[str] = field(default_factory=list)


def index_status(cfg: Settings = settings) -> dict[str, Any]:
    exists = cfg.db_path.is_file() and index_files_exist(cfg)
    if not exists:
        return {"ready": False, "stale": True, "reason": "索引不存在"}
    conn = connect(cfg)
    try:
        meta = get_meta(conn)
    finally:
        conn.close()
    current_hash = data_hash(cfg)
    stale = meta.get("data_hash") != current_hash
    if meta.get("embedding_model") not in {cfg.embedding_model, "hash-fallback"}:
        stale = True
    return {"ready": True, "stale": stale, "meta": meta, "reason": "索引过期" if stale else ""}


class Retriever:
    def __init__(self, cfg: Settings = settings, runtime=None):
        self.cfg = cfg
        self.runtime = runtime
        self.embedder = EmbeddingClient(cfg)
        self.models = ModelClient(cfg)
        self._faiss_index = None
        self._vector_docs: list[dict] | None = None

    def _rt(self, name: str, default):
        if self.runtime is None:
            return default
        return getattr(self.runtime, name, default)

    def search(self, query: str, *, top_k: int | None = None) -> tuple[list[dict], dict[str, float]]:
        top_k = top_k or self._rt("top_k", self.cfg.rerank_top_n)
        timing: dict[str, float] = {}
        t_total = time.perf_counter()

        vector_hits, vec_timing = self._vector_search(query, self.cfg.vector_top_k)
        timing.update(vec_timing)

        t0 = time.perf_counter()
        keyword_hits = self._keyword_search(query, self.cfg.keyword_top_k)
        timing["keyword_search_ms"] = ms_since(t0)

        t0 = time.perf_counter()
        candidates = self._fuse_and_group(vector_hits, keyword_hits)
        timing["fusion_ms"] = ms_since(t0)

        if not self._rt("use_rerank", True):
            candidates.sort(key=lambda c: c.rrf_score, reverse=True)
            timing["rerank_ms"] = 0.0
            results = [self._candidate_to_result(c) for c in candidates[:top_k]]
        else:
            t0 = time.perf_counter()
            ranked = self._rerank(query, candidates, top_k=max(top_k, self.cfg.rerank_top_n))
            timing["rerank_ms"] = ms_since(t0)
            results = [self._candidate_to_result(c) for c in ranked[:top_k]]

        timing["search_ms"] = ms_since(t_total)
        timing["total_ms"] = timing["search_ms"]
        return results, timing

    def chat(self, query: str, *, top_n: int | None = None, use_llm_answer: bool | None = None) -> dict:
        t_total = time.perf_counter()
        top_n = top_n or self._rt("top_n", self.cfg.answer_top_n)
        results, timing = self.search(query, top_k=max(top_n, self.cfg.rerank_top_n))
        min_conf = self._rt("min_confidence_score", self.cfg.min_confidence_score)
        high_conf = bool(results) and float(results[0].get("rerank_score") or results[0].get("rrf_score") or 0) >= min_conf
        if not high_conf:
            timing["generate_ms"] = 0.0
            timing["total_ms"] = ms_since(t_total)
            return {
                "answer": "未找到高置信答案。你可以换一种问法，或查看下方候选结果。",
                "confidence": 0.0,
                "sources": results[:top_n],
                "images": [],
                "mode": "no_high_confidence",
                "timing": timing,
            }
        sources = results[:top_n]
        rt_mode = self._rt("answer_mode", self.cfg.answer_mode)
        should_generate = rt_mode == "generated" or bool(use_llm_answer)
        if should_generate:
            t0 = time.perf_counter()
            answer = self.models.generate_answer(query, sources, runtime=self.runtime)
            timing["generate_ms"] = ms_since(t0)
            mode = "generated"
        else:
            answer = sources[0]["answer"]
            timing["generate_ms"] = 0.0
            mode = "direct"
        images = self._dedupe_images(sources)
        timing["total_ms"] = ms_since(t_total)
        return {
            "answer": answer,
            "confidence": float(sources[0].get("rerank_score") or sources[0].get("rrf_score") or 0),
            "sources": sources,
            "images": images,
            "mode": mode,
            "timing": timing,
        }

    def _load_vector(self):
        if self._faiss_index is not None and self._vector_docs is not None:
            return self._faiss_index, self._vector_docs
        index_path = self.cfg.vector_dir / "faiss.index"
        docs_path = self.cfg.vector_dir / "vector_docs.json"
        self._faiss_index = faiss.read_index(str(index_path))
        self._vector_docs = json.loads(docs_path.read_text(encoding="utf-8"))
        return self._faiss_index, self._vector_docs

    def _vector_search(self, query: str, limit: int) -> tuple[list[dict], dict[str, float]]:
        timing: dict[str, float] = {}
        index, docs = self._load_vector()
        t0 = time.perf_counter()
        q = np.asarray([self.embedder.embed_query(query)], dtype="float32")
        timing["embedding_ms"] = ms_since(t0)
        t0 = time.perf_counter()
        scores, ids = index.search(q, min(limit, len(docs)))
        timing["vector_lookup_ms"] = ms_since(t0)
        out: list[dict] = []
        conn = connect(self.cfg)
        try:
            for score, row_idx in zip(scores[0], ids[0]):
                if row_idx < 0 or row_idx >= len(docs):
                    continue
                doc_id = docs[int(row_idx)]["doc_id"]
                row = conn.execute("SELECT * FROM search_docs WHERE doc_id = ?", (doc_id,)).fetchone()
                if row:
                    item = dict(row)
                    item["score"] = float(score)
                    item["rank_source"] = "vector"
                    out.append(item)
        finally:
            conn.close()
        timing["vector_search_ms"] = round(timing["embedding_ms"] + timing["vector_lookup_ms"], 1)
        return out, timing

    def _keyword_search(self, query: str, limit: int) -> list[dict]:
        q_tokens = list(dict.fromkeys(char_ngrams(query, min_n=2, max_n=3)))[:48]
        conn = connect(self.cfg)
        try:
            rows: list[sqlite3.Row] = []
            if q_tokens:
                fts_query = " OR ".join(q_tokens)
                try:
                    rows = conn.execute(
                        """
                        SELECT s.*, bm25(search_docs_fts) AS rank
                        FROM search_docs_fts
                        JOIN search_docs s ON s.doc_id = search_docs_fts.doc_id
                        WHERE search_docs_fts MATCH ?
                        ORDER BY rank
                        LIMIT ?
                        """,
                        (fts_query, limit),
                    ).fetchall()
                except Exception:
                    rows = []
            if rows:
                return [
                    {
                        **dict(row),
                        "score": 1.0 / (rank + 1),
                        "rank_source": "keyword",
                    }
                    for rank, row in enumerate(rows)
                ]
            return self._keyword_fallback(conn, query, limit)
        finally:
            conn.close()

    def _keyword_fallback(self, conn: sqlite3.Connection, query: str, limit: int) -> list[dict]:
        q_tokens = set(char_ngrams(query, min_n=2, max_n=3))
        rows = conn.execute(
            "SELECT * FROM search_docs WHERE is_eval_holdout = 0"
        ).fetchall()
        scored: list[tuple[float, sqlite3.Row]] = []
        for row in rows:
            text = str(row["keyword_text"] or "")
            score = sum(1 for tok in q_tokens if tok and tok in text)
            if score:
                scored.append((float(score), row))
        scored.sort(key=lambda item: item[0], reverse=True)
        return [
            {**dict(row), "score": score, "rank_source": "keyword"}
            for score, row in scored[:limit]
        ]

    def _fuse_and_group(self, vector_hits: list[dict], keyword_hits: list[dict]) -> list[ParentCandidate]:
        by_item: dict[str, ParentCandidate] = {}

        def add(hit: dict, rank: int, source: str) -> None:
            item_id = str(hit["item_id"])
            cand = by_item.setdefault(item_id, ParentCandidate(item_id=item_id))
            rrf = 1.0 / (self.cfg.rrf_k + rank)
            cand.rrf_score += rrf
            cand.matched_doc_types.add(str(hit["doc_type"]))
            doc_id = str(hit["doc_id"])
            if doc_id not in cand.matched_doc_ids:
                cand.matched_doc_ids.append(doc_id)
            score = float(hit.get("score") or 0.0)
            if source == "vector":
                cand.vector_score = max(cand.vector_score, score)
            else:
                cand.keyword_score = max(cand.keyword_score, score)

        for rank, hit in enumerate(vector_hits, 1):
            add(hit, rank, "vector")
        for rank, hit in enumerate(keyword_hits, 1):
            add(hit, rank, "keyword")
        out = list(by_item.values())
        out.sort(key=lambda c: c.rrf_score, reverse=True)
        return out

    def _rerank(self, query: str, candidates: list[ParentCandidate], top_k: int) -> list[ParentCandidate]:
        conn = connect(self.cfg)
        docs: list[str] = []
        kept: list[ParentCandidate] = []
        try:
            for cand in candidates[: max(top_k * 3, top_k)]:
                item = fetch_item(conn, cand.item_id)
                if not item:
                    continue
                variants = "；".join(v["variant"] for v in item.get("variants", [])[:8])
                docs.append(
                    f"主问题：{item['question']}\n相似问法：{variants}\n答案摘要：{item['answer_summary']}"
                )
                kept.append(cand)
        finally:
            conn.close()
        if not kept:
            return []
        ranked = self.models.rerank(query, docs, top_n=min(top_k, len(kept)))
        output: list[ParentCandidate] = []
        used: set[int] = set()
        for idx, score in ranked:
            if idx < 0 or idx >= len(kept) or idx in used:
                continue
            kept[idx].rerank_score = float(score)
            output.append(kept[idx])
            used.add(idx)
        if len(output) < min(top_k, len(kept)):
            for idx, cand in enumerate(kept):
                if idx not in used:
                    cand.rerank_score = cand.rrf_score
                    output.append(cand)
                if len(output) >= top_k:
                    break
        return output

    def _candidate_to_result(self, cand: ParentCandidate) -> dict:
        conn = connect(self.cfg)
        try:
            item = fetch_item(conn, cand.item_id)
        finally:
            conn.close()
        if not item:
            return {"id": cand.item_id}
        return {
            "id": item["id"],
            "question": item["question"],
            "answer": item["answer"],
            "answer_summary": item["answer_summary"],
            "updated_at": item["updated_at"],
            "images": item["images"],
            "variants": item["variants"],
            "vector_score": cand.vector_score,
            "keyword_score": cand.keyword_score,
            "rrf_score": cand.rrf_score,
            "rerank_score": cand.rerank_score,
            "matched_doc_types": sorted(cand.matched_doc_types),
            "matched_doc_ids": cand.matched_doc_ids,
        }

    @staticmethod
    def _dedupe_images(sources: list[dict]) -> list[dict]:
        seen: set[str] = set()
        images: list[dict] = []
        for src in sources:
            for image in src.get("images") or []:
                key = image.get("src") or image.get("url") or ""
                if key and key not in seen:
                    seen.add(key)
                    images.append(image)
        return images
